src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +74,236 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-                            if (File.Exists(filePath))
-                            {
-                                dotnetPath = Path.GetDirectoryName(filePath);
-                                break;
-                            }
-                            else
-                            {
-                                continue;
-                            }
-                        }
-
-                        dotnetPath = dir;
-                        break;
+                        return NativeLibrary.TryLoad(hostfxrAssembly, out var handle) ? handle : IntPtr.Zero;
                     }
                 }
             }
 
-            if (dotnetPath is null)
-            {
-                throw new InvalidOperationException("Could not find the dotnet executable. Is it on the PATH?");
-            }
+            return IntPtr.Zero;
+        }
 
-            string bestSDK = null;
-            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
+        private static string SdkResolutionExceptionMessage(string methodName) => $"Failed to find all versions of .NET Core MSBuild. Call to {methodName}. There may be more details in stderr.";
+        
+        /// <summary>
+        /// Determines the directory location of the SDK accounting for
+        /// global.json and multi-level lookup policy.
+        /// </summary>
+        private static string? GetSdkFromGlobalSettings(string workingDirectory)
+        {
+            string? resolvedSdk = null;
+            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: DotnetPath.Value, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
                 if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
                 {
-                    bestSDK = value;
+                    resolvedSdk = value;
                 }
             });
 
-            if (rc == 0 && bestSDK != null)
+            if (rc != 0)
             {
-                yield return bestSDK;
+                throw new InvalidOperationException(SdkResolutionExceptionMessage(nameof(NativeMethods.hostfxr_resolve_sdk2)));
             }
-            else if (rc != 0)
+
+            return resolvedSdk;
+        }
+
+        private static string ResolveDotnetPath()
+        {
+            string? dotnetPath = GetDotnetPathFromROOT();
+
+            if (string.IsNullOrEmpty(dotnetPath))
             {
-                throw new InvalidOperationException("Failed to find an appropriate version of .NET Core MSBuild. Call to hostfxr_resolve_sdk2 failed. There may be more details in stderr.");
+                string? dotnetExePath = GetCurrentProcessPath();
+                var isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath) 
+                    && Path.GetFileName(dotnetExePath).Equals(ExeName, StringComparison.InvariantCultureIgnoreCase);
```

Comment:
This looks like something that should be platform-specific, right?

Reply:
It's an execution context-specific - if MsBuild.Locator is invoked in the scope of dotnet.exe, isRunFromDotnetExecutable is true.

Reply:
I meant whether it should be IgnoreCase or not should be platform-specific?